cmd/exit_handler.go

```diff
@@ -0,0 +1,312 @@
+
+	<-handler.ch
+
+	for _, listener := range handler.listeners {
+		listener.Close()
+	}
+	for _, callback := range handler.callbacks {
+		callback.Call()
+	}
+	os.Exit(0)
+}
+
+// AcraExitHandler is a common interface for exit handlers of Acra services
+type AcraExitHandler interface {
+	AddExitSignalHandler(signal AcraSignal, priority Priority)
+	AddDeferFunc(deferFunc func(), priority Priority)
+	ExitZero()
+	ExitOne()
+}
+
+// ExitHandler is an implementation of AcraExitHandler
+type ExitHandler struct {
+	deferFunctions []DeferFunction
+	signalHandlers []AcraSignal
+	notification   chan os.Signal
+}
+
+// NewExitHandler is a constructor for ExitHandler
+func NewExitHandler() (*ExitHandler, error) {
+	return &ExitHandler{
+		deferFunctions: nil,
+		signalHandlers: nil,
+		notification:   make(chan os.Signal),
+	}, nil
+}
+
+// NewSigTERMHandler is a constructor for SigTERMHandler
+func NewSigTERMHandler() AcraSignal {
+	return &SigTERMHandler{
+		nil,
+		nil,
+	}
+}
+
+// NewSigINTHandler is a constructor for SigINTHandler
+func NewSigINTHandler() AcraSignal {
+	return &SigINTHandler{
+		NewSigTERMHandler().(*SigTERMHandler),
+	}
+}
+
+// AddExitSignalHandler appends new signal for handling
+func (s *ExitHandler) AddExitSignalHandler(input AcraSignal) {
+	s.signalHandlers = append(s.signalHandlers, input)
+}
+
+func (s *ExitHandler) waitForExitSystemSignal(exitFuncOnSigINT func(), exitFuncOnSigTERM func()) {
+	signal.Notify(s.notification, s.getSignals()...)
+
+	switch <-s.notification {
+	case syscall.SIGINT:
+		exitFuncOnSigINT()
+	case syscall.SIGTERM:
+		exitFuncOnSigTERM()
+	}
+}
+
+// WaitForExitSystemSignal blocks and waits for signals.
+// It should be used in separate goroutine in main function of service
+func (s *ExitHandler) WaitForExitSystemSignal() {
+	s.waitForExitSystemSignal(s.ExitZero, s.ExitZero)
+}
+
+func (s *ExitHandler) getSignals() []os.Signal {
+	var result []os.Signal
+	for _, signal := range s.signalHandlers {
+		result = append(result, signal.GetSignal())
+	}
+	return result
+}
+
+// AddDeferFunc appends new defer function (with priority) for execution
+func (s *ExitHandler) AddDeferFunc(input DeferFunction) {
+	inputHasLastPriority := input.GetPriority() == Last
+	for _, deferFunc := range s.deferFunctions {
+		if deferFunc.GetPriority() == Last && inputHasLastPriority {
+			panic("defer function with 'Last' priority has been already specified")
+		}
+	}
+	s.deferFunctions = append(s.deferFunctions, input)
+}
+
+// ExitZero is a single point for exiting from the service with 0 code
+func (s *ExitHandler) ExitZero() {
+	s.gracefulExit()
+	os.Exit(0)
+}
+
+// ExitOne is a single point for exiting from the service with 1 code
+func (s *ExitHandler) ExitOne() {
+	s.gracefulExit()
+	os.Exit(1)
+}
+
+func (s *ExitHandler) gracefulExit() {
+	// at first we finalize our handlers (close listeners and call callbacks)
+	s.executeFinalizeOnHandlers()
+	// finally we call defer functions
+	s.executeDeferFunctions()
```

Comment:
It seems that finalization callbacks can be expressed as deferred callbacks. They aren't really that different. Their effect can be achieved with additional priority – `First` – for deferred callbacks. So _first_ we close the listeners, then do whatever additional actions we need, and then _at last_ we sign off on the audit log.

Similarly, we may probably get away with keeping the listener list explicitly. It can be replaced with a bunch of callbacks each closing its individual listener. Sure, keeping an array is more efficient but I doubt it matters that much. However, not making a special case out of the listeners should remove some duplication.

Reply:
Yes. That is true. But we accustomed to use `defer` often. Probably it will be more natural to use `AddDeferFunction` instead of `AddCallback` to termination handler